src/Microsoft.ML.Tokenizers/Model/BPE.cs

```diff
@@ -327,80 +317,124 @@ internal string CharToString(char c)
-        internal Word MergeWord(string w)
+        internal Word MergeWord(ReadOnlySpan<char> w)
         {
             Word word = Word.WithCapacity(w.Length);
             (int Id, int Len)? unk = null;
             int i = 0;
 
-            while (i < w.Length)
-            {
-                int length;
-                string s;
-
-                if (Char.IsHighSurrogate(w[i]) && i < w.Length - 1 && Char.IsLowSurrogate(w[i + 1]))
-                {
-                    length = 2;
-                    s = w.Substring(i, length);
-                }
-                else
-                {
-                    length = 1;
-                    s = CharToString(w[i]);
-                }
+            char[]? buffer = null;
 
-                // Add the `continuing_subword_prefix` if relevant
-                if (i > 0 && ContinuingSubwordPrefix is not null)
+            try
+            {
+                while (i < w.Length)
                 {
-                    s = $"{ContinuingSubwordPrefix}{s}";
-                }
+                    int length;
+                    ReadOnlySpan<char> s;
 
-                // Add the `end_of_word_suffix` if relevant
-                if (i + length >= w.Length && EndOfWordSuffix is not null)
-                {
-                    s = $"{s}{EndOfWordSuffix}";
-                }
+                    if (Char.IsHighSurrogate(w[i]) && i < w.Length - 1 && Char.IsLowSurrogate(w[i + 1]))
+                    {
+                        length = 2;
+                        s = w.Slice(i, 2);
+                    }
+                    else
+                    {
+                        length = 1;
+                        s = w.Slice(i, 1);
+                    }
 
-                if (_vocab.TryGetValue(s, out int id))
-                {
-                    if (unk.HasValue)
+                    // Add the `continuing_subword_prefix` if relevant
+                    if (i > 0 && ContinuingSubwordPrefix is not null)
                     {
-                        word.Add(unk.Value.Id, unk.Value.Len);
-                        unk = null;
+                        if (buffer is null)
+                        {
+                            // 60 should be enough for most cases
+                            buffer = ArrayPool<char>.Shared.Rent(60);
+                        }
+
+                        if (ContinuingSubwordPrefix.Length + s.Length <= buffer.Length)
+                        {
+                            ContinuingSubwordPrefix.AsSpan().CopyTo(buffer.AsSpan());
+                            s.CopyTo(buffer.AsSpan().Slice(ContinuingSubwordPrefix.Length));
+                            s = buffer.AsSpan().Slice(0, ContinuingSubwordPrefix.Length + s.Length);
+                        }
+                        else
+                        {
+                            string s1 = s.Length == 1 ? CharToString(s[0]) : s.ToString();
+                            s = $"{ContinuingSubwordPrefix}{s1}".AsSpan();
```

Comment:
When targeting net6.0+, this would be better as:
```C#
s = $"{ContinuingSubwordPrefix}{s}";
```
That will avoid the intermediate string allocation for s1.